analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExpectedExceptionAttributeShouldNotBeUsed.cs

```diff
@@ -18,15 +18,44 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-namespace SonarAnalyzer.Rules.VisualBasic
+namespace SonarAnalyzer.Rules.VisualBasic;
+
+[DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+public sealed class ExpectedExceptionAttributeShouldNotBeUsed : ExpectedExceptionAttributeShouldNotBeUsedBase<SyntaxKind>
 {
-    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
-    public sealed class ExpectedExceptionAttributeShouldNotBeUsed : ExpectedExceptionAttributeShouldNotBeUsedBase<SyntaxKind>
+    protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+
+    protected override bool HasMultiLineBody(SyntaxNode node) =>
+        node.Parent is MethodBlockSyntax { Statements.Count: > 1 };
+
+    protected override bool AssertInCatchFinallyBlock(SyntaxNode node)
+    {
+        var walker = new CatchFinallyAssertion();
+        foreach (var x in node.Parent.DescendantNodes(_ => true).Where(x => x.IsAnyKind(SyntaxKind.CatchBlock, SyntaxKind.FinallyBlock)))
+        {
+            walker.SafeVisit(x);
+        }
+        return walker.HasAssertion;
+    }
+
+    private sealed class CatchFinallyAssertion : SafeVisualBasicSyntaxWalker
     {
-        protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+        public bool HasAssertion { get; set; }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+        {
+            HasAssertion = node.Expression
+                .ToString()
+                .SplitCamelCaseToWords()
+                .Intersect(UnitTestHelper.KnownAssertionMethodParts)
+                .Any();
+
+            if (HasAssertion)
```

Comment:
This can create FPs, e.g. if a `catch` block has two methods inside: 
```cs
catch (InvalidOperationException exception)
{
    Assert.IsEqual(42, result);
    Console.WriteLine("Hello!");
}
```

To fix this move the `if` statement to the beginning of the method.

Reply:
Why?

I don't think it would create an FP. I will add a test case to make sure.

I am testing the current Invocation, if it is an assertion, I don't want to continue the visit.
Or I don't understand how the walker is working.

Reply:
I was thinking within the same invocation SafeVisit.
It is indeed an issue when calling multiple times SafeVisit on different blocks.

Reply:
It also fixes the FP, I forgot!